site/app/libraries/routers/WebRouter.php

```diff
@@ -84,7 +82,49 @@ public function __construct(Request $request, Core $core, $logged_in, $is_api =
         }
     }
 
+    /**
+     * @param Request $request
+     * @param Core $core
+     * @param $logged_in
+     * @return Response|mixed should be of type Response only in the future
+     */
+    static public function getApiResponse(Request $request, Core $core, $logged_in) {
+        try {
+            $router = new self($request, $core, $logged_in, true);
+        }
+        catch (ResourceNotFoundException $e) {
+            return new Response(JsonResponse::getFailResponse("Endpoint not found."));
+        }
+        catch (AuthenticationException $e) {
+            return new Response(JsonResponse::getFailResponse($e->getMessage()));
+        }
+        catch (MethodNotAllowedException $e) {
+            return new Response(JsonResponse::getFailResponse("Method not allowed."));
+        }
+        catch (\Exception $e) {
+            return new Response(JsonResponse::getErrorResponse($e->getMessage()));
+        }
+
+        $core->getOutput()->disableRender();
+        $core->disableRedirects();
+        return $router->run();
+    }
+
     public function run() {
+        if (!$this->is_api &&
+            $this->request->isMethod('POST') &&
+            !Utils::endsWith($this->parameters['_controller'], 'AuthenticationController') &&
+            !$this->core->checkCsrfToken()
+        ) {
+            $msg = "Invalid CSRF token.";
+            $this->core->addErrorMessage($msg);
+            return new Response(
+                JsonResponse::getFailResponse($msg),
+                null,
+                new RedirectResponse($this->core->buildNewUrl())
+            );
+        }
```

Comment:
I'm not sure we do a redirect on any and all CSRF token failures?

Reply:
Sometimes in our current logic the page is redirected to the previous one. However, I assume that it is a very rare case where the website is running correctly and CSRF token check fails. I assume that this only happens when someone tries to attack Submitty. Therefore it looks okay to me to always redirect home.

Not sure if my assumptions make sense.

Reply:
Yeah, I don't think this particular failure happens every often in the real world, that it's not really worth worrying about the redirect in and of itself. I'm not sure we really handle more common form failures all that gracefully anyway that there's not a good way to really address what should happen here anyway.